### PR TITLE
fix: Reset streams on BFCache events

### PR DESCRIPTION
**Pull request by jiexi**
## **Description**

Previously, Chrome's BFCache (Back Forward) strategy was to evict a page from cache if it received a message over any connected port streams. The port stream would remain open and the background script would NOT receive an onDisconnect. As long as the cached page did not receive a message, the port would still function when the page became active again. This was problematic because MetaMask was likely to send a message to the still connected cached page at some point due to the nature of notifications, which would evict the page from cache, neutralizing the performance benefit of the BFCache for the end user.

Now, Chrome's BFCache strategy is to trigger an onDisconnect for the background script, but NOT the cached page. The port stream is invalid despite not being closed on the cached page side. This is problematic because we do not listen for events relevant to when a BFCached page becomes active and thus do not reset the invalid stream.

To address both strategies, we now listen for the `pageshow` and `pagehide` events. When a page is entering a BFCached state, we preemptively end the port stream connection (even if the user is on an older version of chrome that would have kept it alive). When a BFCached page is restored to an active state, we establish a port stream connection. We know the port stream must be restored/reset because we were the ones responsible for preemptively ending it earlier in the lifecycle. Combining these two changes allows us to handle both the old and new BFCache strategies without having to target them by versions separately. 

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/24950?quickstart=1)

## **Related issues**

See: https://developer.chrome.com/blog/bfcache-extension-messaging-changes?hl=en
See: https://github.com/MetaMask/metamask-extension/issues/13373
See: https://web.dev/articles/bfcache (useful links at bottom)
See: https://github.com/w3c/webextensions/issues/474
Fixes: https://github.com/MetaMask/MetaMask-planning/issues/2582

## **Manual testing steps**
Steps are for macOS. Using chrome 123 or newer

**Testing with old BFCache strategy**
1. Close the entire chrome app
1. run `open /Applications/Google\ Chrome.app --args --disable-features=DisconnectExtensionMessagePortWhenPageEntersBFCache`
1. Visit `http://www.brainjar.com/java/host/test.html`
1. Open console
1. Enter `await window.ethereum.request({method: 'eth_chainId'})`, which should be responsive
1. Visit `chrome://terms/`
1. Use the back button to go back to the brainjar test page
1. Enter `await window.ethereum.request({method: 'eth_chainId'})`, which should be responsive

**Testing with the new BFCache strategy**
Repeat the steps above, but use `--enable-features` instead of `disable`
 
MetaMask Behavior should look the same regardless of browser's BFCache strategy

## **Screenshots/Recordings**

BFCache Behavior

https://github.com/MetaMask/metamask-extension/assets/918701/efeb1591-5fde-44c8-b0a3-3573dfb97806

Prerender Behavior (to show affected chromium browsers still reset streams correctly)

https://github.com/MetaMask/metamask-extension/assets/918701/7461bf64-b5b0-4e70-96d5-416cf5bf6b7c

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: app/scripts/streams/provider-stream.ts [hunk 1]
```diff
@@ -226,19 +226,33 @@ const onMessageSetUpExtensionStreams = (msg: MessageType) => {
...
+export function destroyStreams() {
+  if (!extensionPort) {
+    return;
+  }
+  extensionPort.onDisconnect.removeListener(onDisconnectDestroyStreams);
+
+  destroyExtensionStreams();
+  destroyLegacyExtensionStreams();
+
+  extensionPort.disconnect();
```
Comment (davidmurdoch) [inline app/scripts/streams/provider-stream.ts:242]: Should `METAMASK_EXTENSION_CONNECT_SENT` be set to `false` here?

Comment (jiexi) [reply app/scripts/streams/provider-stream.ts:242]: I don't think that's necessary. If we destroy the streams, no messages can trigger 
[extensionStreamMessageListener](https://github.com/MetaMask/metamask-extension/blob/194b1558e65c0ac65b370adb6bc419e4c705de8b/app/scripts/streams/provider-stream.ts#L312-L333) to fire `METAMASK_EXTENSION_CONNECT_CAN_RETRY` in the first place, but at the same time that means it shouldn't hurt to do so

Comment (jiexi) [reply app/scripts/streams/provider-stream.ts:242]: I'll add it

File: test/e2e/provider/bfcache.spec.js [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+  await driver.executeScript(`
+    window.addEventListener('pageshow', (event) => {
+      if (event.persisted) {
+        window.restoredFromBFCache = true
+      }
+    });
+  `);
+
+  let attempts = 0;
+  while (attempts < 10) {
```
Comment (davidmurdoch) [inline test/e2e/provider/bfcache.spec.js:20]: why do we need multiple attempts, and why `10` and not some other number?

Comment (jiexi) [reply test/e2e/provider/bfcache.spec.js:20]: A page doesn't get BFCached for me consistently. It usually works on the first attempt though. What if we bring this down to 3?

**Review by HowardBraham — DISMISSED** [review]
LGTM.  I also ran the new E2E test on the `develop` version of the Extension, and it failed as expected.

File: test/e2e/provider/bfcache.spec.js (hunk 2 above)
Comment (jiexi) [reply test/e2e/provider/bfcache.spec.js]: looks like i'm wrong and we don't need this anymore. Removed loop here https://github.com/MetaMask/metamask-extension/pull/24950/commits/2dfb287d59490312c700cd4f45f5bf57040462d8

**Review by HowardBraham — APPROVED** [review]

**Review by davidmurdoch — APPROVED** [review]

**Pull request merged**